experimental/sql/dialects/rel_tree.py

```diff
@@ -0,0 +1,223 @@
+
+from ast import Param
+from dataclasses import dataclass
+from typing import Any, List
+from xdsl.ir import Block, Region, Operation, SSAValue, ParametrizedAttribute, Data, MLContext, Attribute
+from xdsl.dialects.builtin import StringAttr, ArrayAttr, ArrayOfConstraint, IntegerAttr
+from xdsl.irdl import AttributeDef, OperandDef, ResultDef, RegionDef, SingleBlockRegionDef, irdl_attr_definition, irdl_op_definition, ParameterDef, AnyAttr, VarOperandDef, builder
+
+# This file contains the relational tree dialect. This dialect represents
+# relational queries in a tree form, simplifying certain optimizations.
+
+
+@irdl_attr_definition
+class DataType(ParametrizedAttribute):
+  """
+  Models a datatype in a relational query.
+  """
+  name = "rel_tree.datatype"
+
+
+@irdl_attr_definition
+class Int32(DataType):
+  """
+  Models a int32 type in a relational query.
+
+  Example:
+
+  ```
+  !rel_tree.int32
+  ```
+  """
+  name = "rel_tree.int32"
+
+
+@irdl_attr_definition
+class String(DataType):
+  """
+  Models a string type in a relational query, that can be either nullable or
+  not.
+
+  Example:
+
+  ```
+  !rel_tree.string<0: !i1>
+  ```
+  """
+  name = "rel_tree.string"
+
+  nullable = ParameterDef(IntegerAttr)
+
+  @staticmethod
+  @builder
+  def get(val: int) -> 'String':
+    return String([IntegerAttr.from_int_and_width(val, 1)])
+
+
+@irdl_op_definition
+class Column(Operation):
+  """
+  References a specific column with name `col_name`.
+
+  Example:
+
+  ```
+  rel_tree.column() ["col_name" = "a"]
+  ```
+  """
+  name = "rel_tree.column"
+
+  col_name = AttributeDef(StringAttr)
+
+  @builder
+  @staticmethod
+  def get(table: Region, col_name: str) -> 'Column':
+    return Column.build(attributes={"col_name": StringAttr.from_str(col_name)},
+                        regions=[table])
+
+
+@irdl_op_definition
+class Predicate(Operation):
+  """
+  Applies the `comparator` to `left` and `right`.
+
+  Example:
+
+  ```
+  rel_tree.predicate() ["comparator" = "="] {
+    rel_tree.column() ...
+  } {
+    rel_tree.literal() ...
+  }
+  ```
+  """
+  name = "rel_tree.predicate"
+
+  comparator = AttributeDef(StringAttr)
+  left = SingleBlockRegionDef()
+  right = SingleBlockRegionDef()
+
+  @builder
+  @staticmethod
+  def get(comparator: str, left: Region, right: Region) -> 'Predicate':
+    return Predicate.build(
+        attributes={"comparator": StringAttr.from_str(comparator)},
+        regions=[left, right])
+
+
+@irdl_op_definition
+class Selection(Operation):
+  """
+  Selects all tuples from `table` that fulfill `predicates`.
+
+  Example:
+
+  ```
+  rel_tree.selection() {
+    rel_tree.pandas_table() ...
+  } {
+    rel_tree.predicate() ...
+  }
+  ```
+  """
+  name = "rel_tree.selection"
+
+  table = SingleBlockRegionDef()
+  predicates = SingleBlockRegionDef()
+
+  @staticmethod
+  @builder
+  def get(table: Region, predicates: Region) -> 'Selection':
+    return Selection.build(regions=[table, predicates])
```

Comment:
In my vision of the relational tree dialect, this predicate/these predicates *including all its/their child expressions* only operate on scalars (as opposed to relations). Maybe it makes sense to introduce two different base classes for this purpose: `RelationlExpression`, which operates on relations (such as selection, project, join, etc.) and `ScalarExpression`, which are used to parametrize some of the former (such as the predicates).

Reply:
I agree with this destinction, but I struggle to see how to implement this in the rel_tree dialect as these things feel like types to me and we don't really have types in the rel_tree (other than data types). This destinction might be something we should make explicit in the dialect below this one.